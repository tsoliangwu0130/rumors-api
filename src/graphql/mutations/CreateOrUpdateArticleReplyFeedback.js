import {
  GraphQLString,
  GraphQLNonNull,
  GraphQLInt,
  GraphQLObjectType,
} from 'graphql';

import { assertUser } from 'graphql/util';
import FeedbackVote from 'graphql/models/FeedbackVote';

import client from 'util/client';

export function getArticleReplyFeedbackId({
  articleId,
  replyId,
  userId,
  appId,
}) {
  return `${articleId}__${replyId}__${userId}__${appId}`;
}

export default {
  description: 'Create or update a feedback on an article-reply connection',
  type: new GraphQLObjectType({
    name: 'CreateOrUpdateArticleReplyFeedbackResult',
    fields: {
      feedbackCount: { type: GraphQLInt },
      positiveFeedbackCount: { type: GraphQLInt },
      negativeFeedbackCount: { type: GraphQLInt },
    },
  }),
  args: {
    articleId: { type: new GraphQLNonNull(GraphQLString) },
    replyId: { type: new GraphQLNonNull(GraphQLString) },
    vote: { type: new GraphQLNonNull(FeedbackVote) },
  },
  async resolve(
    rootValue,
    { articleId, replyId, vote },
    { appId, userId, loaders }
  ) {
    assertUser({ appId, userId });

    const now = new Date().toISOString();

    // (articleId, replyId, userId, appId) should be unique
    // but user can update
    const id = getArticleReplyFeedbackId({
      articleId,
      replyId,
      userId,
      appId,
    });

    await client.update({
      index: 'articlereplyfeedbacks',
      type: 'doc',
      id,
      body: {
        doc: {
          score: vote,
          updatedAt: now,
        },
        upsert: {
          articleId,
          replyId,
          userId,
          appId,
          score: vote,
          createdAt: now,
          updatedAt: now,
        },
      },
      refresh: true, // We are searching for articlereplyfeedbacks immediately
    });

    const feedbacks = await loaders.articleReplyFeedbacksLoader.load({
      articleId,
      replyId,
    });

    const [positiveFeedbackCount, negativeFeedbackCount] = feedbacks.reduce(
      (agg, { score }) => {
        if (score === 1) {
          agg[0] += 1;
        } else if (score === -1) {
          agg[1] += 1;
        }
        return agg;
      },
      [0, 0]
    );

    const articleReplyUpdateResult = await client.update({
      index: 'articles',
      type: 'doc',
      id: articleId,
      body: {
        script: {
          source: `
            int idx = 0;
            int replyCount = ctx._source.articleReplies.size();
            for(; idx < replyCount; idx += 1) {
              HashMap articleReply = ctx._source.articleReplies.get(idx);
              if( articleReply.get('replyId').equals(params.replyId) ) {
                break;
              }
            }

            if( idx === replyCount ) {
              ctx.op = 'none';
            } else {
              ctx._source.articleReplies.get(idx).put(
                'positiveFeedbackCount', params.positiveFeedbackCount);
              ctx._source.articleReplies.get(idx).put(
                'negativeFeedbackCount', params.negativeFeedbackCount);
            }
          `,
          params: {
            replyId,
            positiveFeedbackCount,
            negativeFeedbackCount,
          },
        },
      },
      _source: true,
    });
    if (articleReplyUpdateResult.result !== 'updated') {
      throw new Error(
        `Cannot article ${articleId}'s feedback count for feedback ID = ${id}`
      );
    }

    return {
      feedbackCount: feedbacks.length,
      negativeFeedbackCount,
      positiveFeedbackCount,
    };
  },
};
