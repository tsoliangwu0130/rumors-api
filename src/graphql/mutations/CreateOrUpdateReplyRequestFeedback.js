import {
  GraphQLString,
  GraphQLNonNull,
  GraphQLInt,
  GraphQLObjectType,
} from 'graphql';

import { assertUser } from 'graphql/util';
import FeedbackVote from 'graphql/models/FeedbackVote';

import client from 'util/client';

export default {
  description: 'Create or update a feedback on a reply request reason',
  type: new GraphQLObjectType({
    name: 'CreateOrUpdateReplyRequestFeedbackResult',
    fields: {
      feedbackCount: { type: GraphQLInt },
      positiveFeedbackCount: { type: GraphQLInt },
      negativeFeedbackCount: { type: GraphQLInt },
    },
  }),
  args: {
    replyRequestId: { type: new GraphQLNonNull(GraphQLString) },
    vote: { type: new GraphQLNonNull(FeedbackVote) },
  },
  async resolve(rootValue, { replyRequestId, vote }, { appId, userId }) {
    assertUser({ appId, userId });

    const now = new Date().toISOString();

    const {
      get: {
        _source: { positiveFeedbackCount, negativeFeedbackCount, feedbacks },
      },
    } = await client.update({
      index: 'replyrequests',
      type: 'doc',
      id: replyRequestId,
      body: {
        script: {
          source: `
            int idx = 0;
            int feedbackCount = ctx._source.feedbacks.size();
            for(; idx < feedbackCount; idx += 1) {
              HashMap feedback = ctx._source.feedbacks.get(idx);
              if(
                feedback.get('userId').equals(params.userId) &&
                feedback.get('appId').equals(params.appId)
              ) {
                break;
              }
            }

            if( idx === feedbackCount ) {
              HashMap newFeedback = new HashMap();

              newFeedback.put('userId', params.userId);
              newFeedback.put('appId', params.appId);
              newFeedback.put('score', params.score);
              newFeedback.put('createdAt', params.now);
              newFeedback.put('updatedAt', params.now);

              ctx._source.feedbacks.add(newFeedback);
            } else {
              ctx._source.feedbacks.get(idx).put('score', params.score);
              ctx._source.feedbacks.get(idx).put('updatedAt', params.now);
            }

            ctx._source.positiveFeedbackCount = ctx._source.feedbacks.stream().filter(
              ar -> ar.get('score').equals(1)
            ).count();
            ctx._source.negativeFeedbackCount = ctx._source.feedbacks.stream().filter(
              ar -> ar.get('score').equals(-1)
            ).count();
          `,
          params: {
            userId,
            appId,
            score: vote,
            now,
          },
          lang: 'painless',
        },
      },
      _source: true,
    });

    return {
      feedbackCount: feedbacks.length,
      negativeFeedbackCount,
      positiveFeedbackCount,
    };
  },
};
