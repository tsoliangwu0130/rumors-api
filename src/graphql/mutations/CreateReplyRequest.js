import {
  GraphQLString,
  GraphQLNonNull,
  GraphQLInt,
  GraphQLObjectType,
  GraphQLEnumType,
} from 'graphql';
import { assertUser } from 'graphql/util';

import client, { processMeta } from 'util/client';

/**
 * @typedef {Object} ReplyRequestParam
 * @property {string} articleId - The article to add reply request to
 * @property {string} userId - The user that submits this request
 * @property {string} appId - The app that the user logged in to
 * @property {string} reason - The reason why the user want to submit this article
 */

/**
 * @param {ReplyRequestParam} params
 * @returns {string} the reply request ID for the givne params
 */
export function getReplyRequestId({ articleId, userId, appId }) {
  return `${articleId}__${userId}__${appId}`;
}

/**
 * @typedef {Object} CreateReplyRequestResult
 * @property {Object} article - The update article instance
 * @property {boolean} isCreated - If the reply request is created
 */

/**
 * Indexes a reply request and increments the replyRequestCount for article
 *
 * @param {ReplyRequestParam} param
 * @returns {CreateReplyRequstResult}
 */
export async function createReplyRequest({
  articleId,
  userId,
  appId,
  reason = '',
}) {
  assertUser({ appId, userId });

  const now = new Date().toISOString();
  const id = getReplyRequestId({ articleId, userId, appId });

  const { result } = await client.update({
    index: 'replyrequests',
    type: 'doc',
    id,
    body: {
      doc: {
        updatedAt: now,
      },
      upsert: {
        articleId,
        userId,
        appId,
        reason,
        feedbacks: [],
        positiveFeedbackCount: 0,
        negativeFeedbackCount: 0,
        createdAt: now,
        updatedAt: now,
      },
    },
  });

  const isCreated = result === 'created';

  const articleUpdateResult = await client.update({
    index: 'articles',
    type: 'doc',
    id: articleId,
    body: {
      script: {
        source: `
          ${isCreated ? 'ctx._source.replyRequestCount += 1;' : ''}
          ctx._source.lastRequestedAt = params.now;
        `,
        params: { now },
      },
    },
    _source: true,
  });

  if (articleUpdateResult.result !== 'updated') {
    throw new Error(`Cannot update article ${articleId}`);
  }

  return {
    article: processMeta({ ...articleUpdateResult.get, _id: id }),
    isCreated,
  };
}

export default {
  description: 'Create a reply request for the given article',
  type: new GraphQLObjectType({
    name: 'CreateReplyRequestResult',
    fields: {
      replyRequestCount: {
        type: GraphQLInt,
        description:
          'Reply request count for the given article after creating the reply request',
      },
      status: {
        type: new GraphQLEnumType({
          name: 'CreateReplyRequstResultStatus',
          values: {
            SUCCESS: {
              value: 'SUCCESS',
              description: 'Successfully inserted a new reply request',
            },
            DUPLICATE: {
              value: 'DUPLICATE',
              description:
                'The user has already requested reply for this article',
            },
          },
        }),
      },
    },
  }),
  args: {
    articleId: { type: new GraphQLNonNull(GraphQLString) },
    reason: {
      type: GraphQLString,
      description: 'The reason why the user want to submit this article',
    },
  },
  async resolve(rootValue, { articleId, reason }, { appId, userId }) {
    const { article, isCreated } = await createReplyRequest({
      articleId,
      appId,
      userId,
      reason,
    });
    return {
      replyRequestCount: article.replyRequestCount,
      status: isCreated ? 'SUCCESS' : 'DUPLICATE',
    };
  },
};
