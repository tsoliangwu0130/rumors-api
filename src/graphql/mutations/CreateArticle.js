import { GraphQLString, GraphQLNonNull } from 'graphql';
import { h64 } from 'xxhashjs';

import { assertUser } from 'graphql/util';
import client from 'util/client';

import { ArticleReferenceInput } from 'graphql/models/ArticleReference';
import MutationResult from 'graphql/models/MutationResult';
import { createReplyRequest } from './CreateReplyRequest';

/* Instantiate hash function */
const xxhash64 = h64();

/**
 * Generates ID from article text. Outputs identical ID when the given the same article text.
 *
 * Used for preventing identical articles sending in within a short amount of time,
 * i.e. while other articles are sending in.
 *
 * @param {string} text The article text
 * @returns {string} generated article ID
 */
export function getArticleId(text) {
  return xxhash64
    .update(text)
    .digest()
    .toString(36);
}

/**
 * Creates a new article in ElasticSearch,
 * or updates its reference by adding a new reference
 *
 * @param {object} param
 * @param {string} param.text
 * @param {ArticleReferenceInput} param.reference
 * @param {string} param.userId
 * @param {string} param.appId
 * @returns {string} the new article's ID
 */
async function createNewArticle({
  text,
  reference: originalReference,
  userId,
  appId,
}) {
  const articleId = getArticleId(text);
  const now = new Date().toISOString();
  const reference = {
    ...originalReference,
    createdAt: now,
    userId: userId,
    appId: appId,
  };

  await client.update({
    index: 'articles',
    type: 'doc',
    id: articleId,
    body: {
      script: {
        source: `
          ctx._source.updatedAt = params.updatedAt;
          ctx._source.references.add(params.reference);
        `,
        lang: 'painless',
        params: {
          updatedAt: now,
          reference,
        },
      },
      upsert: {
        text,
        createdAt: now,
        updatedAt: now,
        userId,
        appId,
        references: [reference],
        articleReplies: [],
        normalArticleReplycount: 0,
        replyRequestCount: 0,
        tags: [],
      },
    },
    refresh: true, // Make sure the data is indexed when we create ReplyRequest
  });

  return articleId;
}

export default {
  type: MutationResult,
  description: 'Create an article and/or a replyRequest',
  args: {
    text: { type: new GraphQLNonNull(GraphQLString) },
    reference: { type: new GraphQLNonNull(ArticleReferenceInput) },
    reason: {
      type: new GraphQLNonNull(GraphQLString),
      description:
        'The reason why the user want to submit this article. Mandatory for 1st sender',
    },
  },
  async resolve(rootValue, { text, reference, reason }, { appId, userId }) {
    assertUser({ appId, userId });

    const articleId = await createNewArticle({
      text,
      reference,
      userId,
      appId,
    });

    await createReplyRequest({ articleId, userId, appId, reason });

    return { id: articleId };
  },
};
