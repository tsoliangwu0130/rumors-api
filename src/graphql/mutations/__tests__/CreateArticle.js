import gql from 'util/GraphQL';
import { loadFixtures, unloadFixtures } from 'util/fixtures';
import client from 'util/client';
import MockDate from 'mockdate';
import fixtures, { fixture1Text } from '../__fixtures__/CreateArticle';
import { getReplyRequestId } from '../CreateReplyRequest';
import { getArticleId } from 'graphql/mutations/CreateArticle';

it('creates articles and a reply request', async () => {
  MockDate.set(1485593157011);
  const userId = 'test';
  const appId = 'foo';

  const { data, errors } = await gql`
    mutation($text: String!, $reference: ArticleReferenceInput!) {
      CreateArticle(
        text: $text
        reference: $reference
        reason: "気になります"
      ) {
        id
      }
    }
  `(
    {
      text: 'FOO FOO',
      reference: { type: 'LINE' },
    },
    { userId, appId }
  );
  MockDate.reset();

  expect(errors).toBeUndefined();

  const { _source: article } = await client.get({
    index: 'articles',
    type: 'doc',
    id: data.CreateArticle.id,
  });

  expect(article.replyRequestCount).toBe(1);
  expect(article).toMatchSnapshot();

  const replyRequestId = getReplyRequestId({
    articleId: data.CreateArticle.id,
    userId,
    appId,
  });

  const { _source: replyRequest } = await client.get({
    index: 'replyrequests',
    type: 'doc',
    id: replyRequestId,
  });

  expect(replyRequest).toMatchSnapshot();

  // Cleanup
  await client.delete({
    index: 'articles',
    type: 'doc',
    id: data.CreateArticle.id,
  });

  await client.delete({
    index: 'replyrequests',
    type: 'doc',
    id: replyRequestId,
  });
});

it('avoids creating duplicated articles and adds replyRequests automatically', async () => {
  await loadFixtures(fixtures);

  MockDate.set(1485593157011);
  const userId = 'test';
  const appId = 'foo';
  const articleId = getArticleId(fixture1Text);

  const { data, errors } = await gql`
    mutation($text: String!, $reference: ArticleReferenceInput!) {
      CreateArticle(
        text: $text
        reference: $reference
        reason: "気になります"
      ) {
        id
      }
    }
  `(
    {
      text: fixture1Text,
      reference: { type: 'LINE' },
    },
    { userId, appId }
  );
  MockDate.reset();
  expect(errors).toBeUndefined();

  // Expects no new article is created,
  // and it returns the existing ID
  expect(data.CreateArticle.id).toBe(articleId);

  const { _source: article } = await client.get({
    index: 'articles',
    type: 'doc',
    id: articleId,
  });

  // Expects lastRequestedAt, references are updated
  expect(article).toMatchSnapshot();

  // Expects new replyRequest is generated
  const replyRequestId = getReplyRequestId({ articleId, appId, userId });
  const { _source: replyRequest } = await client.get({
    index: 'replyrequests',
    type: 'doc',
    id: replyRequestId,
  });

  expect(replyRequest).toMatchSnapshot();

  // Cleanup
  await client.delete({
    index: 'replyrequests',
    type: 'doc',
    id: replyRequestId,
  });

  await unloadFixtures(fixtures);
});
