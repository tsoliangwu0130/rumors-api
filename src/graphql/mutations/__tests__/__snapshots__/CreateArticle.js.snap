// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`avoids creating duplicated articles and adds replyRequests automatically 1`] = `
Object {
  "lastRequestedAt": "2017-01-28T08:45:57.011Z",
  "references": Array [
    Object {
      "type": "LINE",
    },
    Object {
      "appId": "foo",
      "createdAt": "2017-01-28T08:45:57.011Z",
      "type": "LINE",
      "userId": "test",
    },
  ],
  "replyRequestCount": 2,
  "text": "I think I am I exist",
  "updatedAt": "2017-01-28T08:45:57.011Z",
}
`;

exports[`avoids creating duplicated articles and adds replyRequests automatically 2`] = `
Object {
  "appId": "foo",
  "articleId": "20xz7y20qzyt6",
  "createdAt": "2017-01-28T08:45:57.011Z",
  "feedbacks": Array [],
  "negativeFeedbackCount": 0,
  "positiveFeedbackCount": 0,
  "reason": "気になります",
  "updatedAt": "2017-01-28T08:45:57.011Z",
  "userId": "test",
}
`;

exports[`creates articles and a reply request 1`] = `
Object {
  "appId": "foo",
  "articleReplies": Array [],
  "createdAt": "2017-01-28T08:45:57.011Z",
  "lastRequestedAt": "2017-01-28T08:45:57.011Z",
  "normalArticleReplycount": 0,
  "references": Array [
    Object {
      "appId": "foo",
      "createdAt": "2017-01-28T08:45:57.011Z",
      "type": "LINE",
      "userId": "test",
    },
  ],
  "replyRequestCount": 1,
  "tags": Array [],
  "text": "FOO FOO",
  "updatedAt": "2017-01-28T08:45:57.011Z",
  "userId": "test",
}
`;

exports[`creates articles and a reply request 2`] = `
Object {
  "appId": "foo",
  "articleId": "mdw91a2ngnnt",
  "createdAt": "2017-01-28T08:45:57.011Z",
  "feedbacks": Array [],
  "negativeFeedbackCount": 0,
  "positiveFeedbackCount": 0,
  "reason": "気になります",
  "updatedAt": "2017-01-28T08:45:57.011Z",
  "userId": "test",
}
`;
